Updated standardise module, applied to tests

brendan-m-murphy opened the pull request:
This PR refactors tests to use "standardise" functions instead of directly using BaseStore children.
This serves two purposes:
1) It means that we can change how data is standardised without modifying all the tests.
2) It shows points where BaseStore children are (currently) being used to standardise data, 
as opposed to being used to access the metastore.

* **Summary of changes** (Bug fix, feature, docs update, ...)

Previously, `standardise_surface`, etc. could not accept a bucket as an argument, which meant that BaseStore children needed to be used in a context manager to standardise files. Now these functions do accept a bucket.

`standardise_surface`, `standardise_flux`, `standardise_footprints`, etc. now all call a generic `standardise` function that accepts a `data_type` argument.

I also added standardising function for binary data, remote data, and transformed data (these also accept a `data_type` argument). These functions are kind of ugly, but they're meant to help refactor the BaseStore children by providing an interface that hides the actual implementation details.

* **Please check if the PR fulfills these requirements**

- [ ] Closes #xxxx (Replace xxxx with the Github issue number)
- [ ] [Tests added and passed](https://docs.openghg.org/development/python_devel.html#testing) if fixing a bug or adding a new feature
- [ ] All code checks passing - `black --line-length 110` run over code, `mypy` and `flake8` not showing any errors
- [ ] Documentation and tutorials updated/added
- [ ] Added an entry in the latest `CHANGELOG.md` file if fixing a bug or adding a new feature
- [ ] Added any new requirements to `requirements.txt` and `recipes/meta.yaml`

rt17603 commented on openghg/standardise/_standardise.py:
```diff
@@ -489,72 +508,48 @@ def standardise_flux(
+    return result
+
+
+def standardise_from_remote_source(bucket: str, data_type: str, data: dict, **kwargs: Any) -> Optional[dict]:
```
Missing doc string here (and including that would make it easier to understand what this new function is doing)

brendan-m-murphy replied:
I removed this function. It was meant to be a front end for `store_data`, but I've just changed that back to using the `ObsSurface` context manager. (Most likely I will remove that context manager when I made the Metastore class, so I'll have to change the call slightly then, but I guess that is better than having an out of place function in standardise.)

rt17603 commented on openghg/standardise/_standardise.py:
```diff
@@ -489,72 +508,48 @@ def standardise_flux(
+
+
+def standardise_from_binary_data(
+    bucket: str, data_type: str, binary_data: bytes, metadata: dict, file_metadata: dict, **kwargs: Any
```
Missing doc-string

brendan-m-murphy replied:
fixed

rt17603 commented on openghg/standardise/_standardise.py:
```diff
@@ -489,72 +508,48 @@ def standardise_flux(
+    return result
+
+
+def standardise_via_transform(
```
See Issue #588 in relation to the plans for the transform functionality. This is meant to be expanded and sit in parallel to standardise rather than going through the same interface.

brendan-m-murphy replied:
I added a _transform.py module to the top level of openghg.transform and put `transform_emissions_data` in there, which does the same thing that this function did. (I added a doc string as well.)

rt17603 commented on tests/db_integrity/test_data_updates.py:
```diff
@@ -261,14 +252,9 @@ def test_obs_data_read_header_diff():
     """
-    clear_test_stores()
-    # Load BSD data - CRDS data
     bsd_data_read_crds()
```
Please add the clarifying comments be added back here.

brendan-m-murphy replied:
Done

rt17603 commented on tests/db_integrity/test_data_updates.py:
```diff
@@ -511,10 +471,9 @@ def test_obs_data_representative_date_overlap():
-    bsd_data_read_crds_internal_overlap()
-    bsd_data_read_crds_internal_overlap(overwrite=True)
+    # Add same data twice, overwriting the second time
+    bsd_crds_overwrite(False)
```
I feel the previous setup for this was more clear - this was a test for a specific bug we found rather than a generic overwrite of crds data.

brendan-m-murphy replied:
I put the tests with overwrite parameters back as they were

rt17603 commented on tests/store/test_eulerian_model.py:
```diff
@@ -1,16 +1,17 @@
     bucket = get_bucket()
-    with EulerianModel(bucket=bucket) as em:
-        proc_results = em.read_file(filepath=test_datapath, model="GEOSChem", species="ch4")
+    proc_results = standardise(bucket=bucket, data_type="eulerian_model", filepath=test_datapath, model="GEOSChem", species="ch4")
```
What we should do really is add a `standardise_eulerian` function and test this here. I'm adding this a suggestion though rather than saying this needs to be done in this PR (otherwise we should open an issue).

brendan-m-murphy replied:
I've now added this function

rt17603 commented on tests/store/test_footprints.py:
```diff
@@ -121,34 +126,25 @@ def test_read_footprint_standard(keyword, value):
-            domain=domain,
-            period="monthly",
-            high_spatial_resolution=True,
+    standardise(
```
Small comment - any reason why this is laid out differently than the `test_read_footprint_standard`? I like how that test was laid out, so a small suggestion would be to match this to that to keep consistent.

brendan-m-murphy replied:
Partially because I could pass keyword=value through **kwargs, but not directly. I changed it so that only keyword=value is in kwargs.

rt17603 replied:
This isn't a strong suggestion so if you feel like there's a good reason to lay it out how you did that's fine.

rt17603 commented on tests/store/test_footprints.py:
```diff
@@ -277,7 +273,7 @@ def test_read_footprint_high_spatial_resolution():
     ],
 )
-def test_read_footprint_co2(site, inlet, metmodel, start, end, filename):
+def test_read_footprint_co2(site, inlet, metmodel, start, end, filename, standardise):
```
Same small comment as for `test_read_footprint_high_spatial_resolution` - would be nice to keep the variable definition and layout consistent.

rt17603 commented on tests/db_integrity/test_data_updates.py:
```diff
@@ -6,108 +6,80 @@
+    as arguments to the test, so this seems like the best way
+    to do this.
+    """
+    global bucket
```
Can I ask what problem this is aiming to solve? Is the idea to allow `bucket` to be accessible to all tests without having to specify it each time?

I think overall we should avoid global state where we can as it can easily cause confusion. Using a fixture with the functions directly and without the autouse may be a more clear approach here.

brendan-m-murphy replied:
It's the same as calling "bucket" as a fixture each time. I could make it function scope and then it would essentially be a local variable.

I guess it serves the same purpose as the fixture that clears test stores each time.

I could just call the fixture explicitly if you prefer that.

rt17603 replied:
Yes I think if we can avoid global state where we can, that would be better.

rt17603 commented on tests/db_integrity/test_data_updates.py:
```diff
@@ -6,108 +6,80 @@
-def flux_data_read():
-    """
-    Flux data set up.
+@pytest.fixture(autouse=True)
```
I think we should be careful with overusing `autouse` as it can cause confusion - my feel is that for these tests in particular we'd be better off being more explicit with clearing the store at the start of each test to make it clear the steps that have been taken.

rt17603 replied:
@brendan-m-murphy - just a lingering comment from previous review on the use of autouse for clear_stores in test_data_updates.py to address.

brendan-m-murphy commented on tests/db_integrity/test_data_updates.py:
```diff
@@ -141,18 +112,17 @@ def bsd_small_edit_data_read():
 
 
-def bsd_diff_data_read(overwrite=False):
+def bsd_diff_data_read():
```
For the next 3-4 comments, in an earlier version I was trying to use fixtures to help set things up and it ended up being more complicated, so I tried to restore it... I'll try to put it back to closer to what it was.

rt17603 commented on tests/store/test_emissions.py:
```diff
@@ -11,43 +13,67 @@
+    return partial(standardise_flux, bucket=bucket)
+
+
+class TestCardamom:
```
I understand why this approach has been taken to create test classes to group together functionality but if this needs to be split out in this way, this may be better split into two separate files instead, linking with the appropriate fixtures. If possible it would be preferable to keep this as a set of tests in one file, linking to fixtures if different data needs to be set up.

A few other points here:
 - Often the data is used not because it's EDGAR / Cardamom data but as a way to simulate a particular set of steps. So splitting by the two emissions types here doesn't necessarily map to the purpose of the tests.
 - As long as we understand what we're testing, having both EDGAR and Cardamom data within the same object store is not a wholly bad thing (as it could highlight some problems we hadn't anticipated) - however I do appreciate this is a less clear and predictable approach for the tests.

brendan-m-murphy replied:
I think I did this because I figured the groups made sense. I'll just drop the classes if the groups don't make sense. (Some book I have suggested using classes for groups of related tests. They said they don't typically use them for grouping fixtures, but it seemed to make sense in this case.)

rt17603 commented on tests/store/test_footprints.py:
```diff
@@ -1,10 +1,32 @@
+    globals().pop('bucket')
+
+
+@pytest.fixture(autouse=True)
```
In this case, is the autouse producing anything here (based on your previous comments around the reason for global bucket)?

It looks like `standardise` is used within the tests as needed as an input?

brendan-m-murphy replied:
I think you're right, I removed the "auto_bucket" fixture and standardise fixture and just called bucket where needed.

rt17603 commented on tests/store/test_footprints.py:
```diff
@@ -1,10 +1,32 @@
+    This was added to help refactor `BaseStore` and its children
+    by removing uses of the context manager for using `read_file`.`
+    """
+    return partial(standardise_footprint, bucket=bucket)
```
Again, has this been created as a means to reduce repetition for creating the bucket?

brendan-m-murphy replied:
Yes, I didn't figure out how to make an autouse fixture set a variable (e.g. bucket) until after I had updated these tests. And originally the `standardise` fixture did a lot more, because I didn't update the standardise_* functions at first, then later decided it made sense to do that.

rt17603 commented on CHANGELOG.md:
```diff
@@ -9,6 +9,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [0.6.2] - 2023-08-07
```
As 0.6.2 has been released, this should be included under the "Unreleased" section instead.

rt17603 reviewed (changes requested):
Great work - some very useful restructuring and clarity added to the code here with some nice features.

I have some questions about some of the new functions added and around some updates to the tests to be addressed.

brendan-m-murphy commented on tests/db_integrity/test_data_updates.py:
```diff
@@ -164,20 +134,42 @@ def bsd_diff_data_read(overwrite=False):
+    )
+
+
+def bsd_diff_date_range_read_delayed():
```
I removed this... it's a duplicate of one of the original helper functions.

brendan-m-murphy commented:
I think I addressed everything you brought up. Some of the test files were a bit of a mess, I think they're better now (no autouse fixtures, and I replaced "standardise" fixtures with the proper functions from the standardise module). I added a transform module to mirror the standardise module, added docstrings, and got rid of one function in standardise that didn't fit that well.

rt17603 commented on openghg/transform/_transform.py:
In general, adding this sort of parallel functionality should probably be included on a separate branch and PR (and will close the #588 issue mentioned previously).

As it is, if we're keeping this here, we should also update the associated tutorial to now use this new function - https://github.com/openghg/openghg/blob/devel/doc/source/tutorials/local/Adding_new_data/Creating_data_from_databases.rst#1-creating-emissions-from-the-edgar-database

rt17603 commented on tests/db_integrity/test_data_updates.py:
```diff
@@ -511,10 +462,9 @@ def test_obs_data_representative_date_overlap():
+    # Add same data twice, overwriting the second time
     bsd_data_read_crds_internal_overlap()
-    bsd_data_read_crds_internal_overlap(overwrite=True)
+    bsd_data_read_crds_internal_overlap(overwrite=False)
```
This should include `overwrite=True`

rt17603 commented on tests/store/test_emissions.py:
```diff
@@ -108,10 +123,33 @@ def test_read_file():
 
+@pytest.fixture
+def load_edgar(bucket):
+    def _load_edgar(version, year, database_version: bool = True):
```
There's a few assumptions here we can't necessarily make around EDGAR data
 - For the database version there was previously a v4.3.2 and there may be other versions going forward which use similiar syntax
 - We may want to look at monthly edgar data rather than yearly (that's a case we will need to add in future)
 - We may also want to look at a different species e.g. co2 or no2

Overall, I think we'd be better off just allowing this be flexibly called rather than setting this up as a more clever fixture.

rt17603 replied:
And think this isn't an update which is needed in this PR

brendan-m-murphy replied:
This fixture is really only meant for the tests in that file. I added it because otherwise I would have had to make the standardize changes in 10 places instead of 1. I can substitute the body of the fixture back into the tests, but I still need to update those tests one way or another.

rt17603 replied:
Sure, I'm thinking about building in flexibility which will be needed in this test file (other versions / gases for edgar will likely be included in tests within this file).

A few small additions which may address this:
 - make the full string for the version be included in the input (solves the "v4.3.2" issue)
 - add species as an input (e.g. "ch4")

rt17603 reviewed (changes requested):
Thanks for making the changes - most look good, just a lingering comment on autouse to address and a few more comments based on the updates.

You may also consider splitting out some of the added functionality into a different PR (e.g. adding the transform functionality; updating the way edgar is accessed in the test files) but we can leave this for now.

brendan-m-murphy commented:
@rt17603 I made the changes suggested in the second code review. I didn't split off the `transform` module or EDGAR test changes into separate PRs; I can do this if you like.

rt17603 commented on tests/db_integrity/test_data_updates.py:
```diff
@@ -2,112 +2,74 @@
-    """
-    Flux data set up.
+@pytest.fixture
+def clear_stores():
```
One final detail around this - at this point these fixtures have just ended up as function calls. Re-evaluating these changes in this state, it might make more sense to either:
1. Just add the function calls themselves back to the tests rather than using these fixtures
2. Combine these two fixtures to get and then clear the bucket (could then return bucket value to be used within the tests)

brendan-m-murphy replied:
All of the tests clear the stores, so they all need to call `clear_stores`, which is why it was autouse, but not all of them need `bucket` as an argument now, because Gareth suggested that I change the standardise functions to only accept 'store' as an argument, and not 'bucket'.

Gareth is also suggesting that `get_bucket()` be replaced by `get_writeable_buckets(name="user")`, so if I add the function calls back then that change becomes harder to make.

Changing the helper functions to take 'bucket' as an argument instead of inheriting through the `auto_bucket' fixture made making the changes Gareth suggested about 'store' instead of 'bucket' harder to do, so I feel like removing some of these fixtures has already made making changes harder. (I admit the global argument in auto_bucket seemed slightly dodgy, although I don't think anything bad could happen because globals are defined at the point where functions are defined, so it only affects code within the same module, which consists of tests and helper functions.)

So, I don't think either option is great, given the changes suggested by Gareth.

rt17603 replied:
The first option would be the best one to go with then. There's no reason to have `clear_stores` as a fixture if all it contains is one function call and there's no issue really with having that as a function at the start each test.

Clarity is much more important than brevity when it comes to tests (especially this set of tests). The problem with the autouse was that this reduced clarity especially when the purpose of these specific tests was to very clearly check a particular state of the object store.

Overall this could all be improved by creating a "give me an empty bucket" fixture which creates a new empty bucket and then tidies this up afterwards (which is what I was partially suggesting the fixture at the top could be turned into) but we can use what we have to work with for now.

brendan-m-murphy replied:
Okay I will put the code back into the tests and remove the fixtures. The combined fixture wouldn't do any harm, but not all of the tests need to use a bucket, so I don't think that aids clarity.

I personally think the autouse fixture for clearing stores is more explicit in this case, since we want every test to start from a blank state; if the code is moved to the tests, then it's not obvious that all of these tests start from a blank slate (in some test modules, clearing stores isn't done uniformly).

Removing the bucket fixture isn't that big of a deal since I think the change to `get_writeable_buckets` should just be search and replace.

rt17603 replied:
Thanks for applying that change. I do understand what you're saying - you were aiming to get everything set up from a known base state of an empty object store (and I do think one resolution to consider for future would be to have a `get_empty_bucket` type input).

However, overall we should be careful with what is essentially hidden state and autouse functions and global variables can act as that - I've definitely encountered problems with that kind of setup in the past when trying to debug tests and setting up mocking for instance.

rt17603 reviewed (changes requested):
Thanks for addressing comments so far on this - looks good. One last tidy up step for test_data_updates.py to be addressed based on its current state - see comment in file for suggestions here.

rt17603 commented:
> @rt17603 Gareth requested some changes on PR #767, but he meant them to go with this PR, so I've made those changes. Those changes affect the code in your final requested changed, and I don't think either option is very good now.

When asking for review I would always recommend asking at the point where you're satisfied the PR is completed and ready to be looked at. This includes the "Files changed" tab showing the changes specifically related to that PR - that way it's much easier to review.

For chained PRs asking for a review on both means we often end up with changes on both branches and more chance of merge conflicts (and a more complex history). So yes I would recommend requesting one review after the other rather than both at the same time.

rt17603 reviewed (approved):
Looks good - happy for this to be merged.
